Initial event-native documentation (#935)
gravitee-io/gravitee-docs: pages/apim/3.x/v4-beta/v4-beta-event-native-apim-user-guide.adoc

====

[TIP]
====
Before running any of the examples on this page, work through some of the  link:event_native_apim_example_use_cases.html[example use cases] so you know that the event-native engine is running.
====

== Asynchronous APIs

In this release, you can create asynchronous APIs and connect to event brokers and asynchronous data sources. You can configure asynchronous API connections using the RESTful interface to the management API.

The OpenAPI specification for the management API is available at link:/apim/3.x/management-api/3.20/swagger.json[`swagger.json`] (link:/apim/3.x/apim_installguide_rest_apis_documentation.html#apim_console_api_reference[reference documentation]). Examples of how to use the management API to configure asynchronous APIs can be found in the link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee Postman Collections], which are described in link:event_native_apim_example_use_cases.html[example use cases].

[NOTE]
====
The management API is in beta and it may change.
====

== Entrypoints and endpoints

> An _entrypoint_ is how an end user connects to the Gravitee gateway. An _endpoint_ is how the Gravitee gateway connects to the source of data.

Reviewer: ```suggestion
An _entrypoint_ is how an end user connects to the Gravitee Gateway. An _endpoint_ is how the Gravitee Gateway connects to the source of data.
```

> The following types of entrypoint are available.

Reviewer: ```suggestion
The following types of entrypoints are available:
```

> * SSE

Reviewer: ```suggestion
* SSE (Server-Sent Events)
```

> * SSE advanced (EE)

Reviewer: ```suggestion
* SSE advanced (Enterprise Edition only)
```

> * Websockets

Reviewer: ```suggestion
* WebSocket
```

* HTTP GET
* HTTP POST
* Webhook

> * Webhook advanced (EE)

Reviewer: ```suggestion
* Webhook advanced (Enterprise Edition only)
```

> For more information about these entrypoints, see link:{{'/apim/3.x/event_native_apim_introduction.html#http_post_and_http_get_entrypoints' | relative_url}}[HTTP POST and HTTP GET entrypoints] and link:{{'/apim/3.x/event_native_apim_introduction.html#support_for_websocket_and_sse_entrypoints' | relative_url}}[Support for Websocket and SSE entrypoints]

Reviewer: ```suggestion
For more information about these entrypoints, see link:{{'/apim/3.x/event_native_apim_introduction.html#http_post_and_http_get_entrypoints' | relative_url}}[HTTP POST and HTTP GET entrypoints] and link:{{'/apim/3.x/event_native_apim_introduction.html#support_for_websocket_and_sse_entrypoints' | relative_url}}[Support for Websocket and SSE entrypoints].
```

> The following types of endpoint are available

Reviewer: ```suggestion
The following types of endpoints are available:
```

* Kafka

> * Kafka advanced (EE)

Reviewer: ```suggestion
* Kafka advanced (Enterprise Edition only)
```

* MQTT

> * MQTT advanced (EE)

Reviewer: ```suggestion
* MQTT advanced (Enterprise Edition only)
```

> == Create a v4 asynchronous API

Reviewer: ```suggestion
== Create a V4 BETA asynchronous API
```

> The link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee Postman Collections] contain several examples of how to create an API.

Reviewer: ```suggestion
The link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee V4 BETA Postman Collection] contains several examples of how to create and test an asynchronous API using the event-native V4 BETA API definition and link:{{'/apim/3.x/v4_new_policy_execution_engine_introduction.html' | relative_url}}[the new V4 BETA policy execution engine] .
```

            ]
        }
    ],
    "flows": [
        {
            "name": "",
            "selectors": [],
            "request": [],
            "response": [],
            "subscribe": [],
            "publish": [],
            "enabled": true
        }
    ]
}
----

== Quality of Service

When working with asynchronous APIs, quality of service is important. Quality of service defines the guaranteed level of message delivery. For example, a quality of service of "none" means that a given message might be delivered zero, one, or several times. A quality of service of "at-most-once" means that a given message will be delivered zero or one times, with no duplication.

A higher quality of service could lead to lower system performance depending on the endpoint chosen.

The quality-of-service levels are described in the following table.

.Table Quality of service levels
[cols="1,4"]
|===
|Level | Description

| None
| A given message might be delivered zero, one, or many times. This level allows high throughput and good performance but without guaranteed delivery. After failure or disconnection, the client will only receive messages sent after reconnection.

| Auto (0 or N)

> | A given message might be delivered zero, one, or many times. This level allows a trade-off between performance and delivery guarantee. Delivery is highly dependant on the capabilities supported by the endpoint connector.  After failure or disconnection, after reconnection, the client will resume, if possible, from a previously saved state, although messages duplication could potentially exist.

Reviewer: ```suggestion
| A given message might be delivered zero, one, or many times. This level allows a trade-off between performance and delivery guarantee. Delivery is highly dependent on the capabilities supported by the endpoint connector.  In case of failure or disconnection, after reconnection the client will resume, if possible, from a previously saved state, although duplication of messages could potentially exist.
```

| At-Most-Once (0 or 1)
| A given message might be delivered zero times or once without any duplication. Depending on the capabilities of the entrypoint connector, performance could be degraded.

| At-Least-Once (1 or N)
| A given message is delivered once or many times. This level gives a good balance between guaranteed delivery and performance when compared to At-Most-Once, especially when the entrypoint connector is not able to resume message streams after failure.

|===

=== Setting quality of service

You can set quality of service with the `qos` object in the `entrypoints` object, as shown in the following example. See the link:/apim/3.x/management-api/3.20/swagger.json[`swagger.json`] definition of the Management API For values that `qos` can have.

[source json]
----
"entrypoints": [
                {
                    "type": "sse",
                    "qos": "none",
                    "configuration": {
                        "heartbeatIntervalInMs": 5000,
                        "metadataAsComment": false,
                        "headersAsComment": false
                    }
                }
            ]
----

=== Compatibility

> Not all qualities of service work with every combination of entrypoint and endpoint. The following table shows how they can be combined.

Reviewer: ```suggestion
Not all levels of quality of service work with every entrypoint / endpoint combination. The following table shows how they can be used.
```

.Table Quality of service compatibility matrix
|===

> | Entrypoint| MQTT5 | MQTT5 advanced | Kafka | Kafka advanced

Reviewer: ```suggestion
| Entrypoint| MQTT5 endpoint | MQTT5 advanced endpoint | Kafka endpoint | Kafka advanced endpoint
```

| none, auto
| none, auto

| SSE advanced
| none, auto
| none, auto
| none, auto
| none, auto, at least once, at most once

| HTTP POST
| none, auto
| none, auto
| none, auto
| none, auto

| HTTP GET
| auto
| auto
| auto
| auto, at least once, at most once

| Websocket
| none, auto
| none, auto
| none, auto
| none, auto

| SSE advanced
| none, auto
| none, auto, at least once, at most once
| none, auto
| none, auto, at least once, at most once

|===

== Policies

Policies are steps in the gateway execution chain. A policy guarantees that a given business rule will be fulfilled during processing.

Policies can be set on request, response, subscribe, or publish phases. The following example shows how to set a policy on a subscribe phase.

[source json]
----
"subscribe": [
                {
                    "name": "Message filtering",
                    "description": "Apply filter to messages",
                    "enabled": true,
                    "policy": "message-filtering",
                    "configuration": {
                        "filter": "{#message.headers.foo == #subscription.metadata['bar']}"
                    }
                }
            ]
----

> For an example, see _04 - Event Consumption - Webhook_ > _Webhook Messaging Filtering_ > _Create API_ in The link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee Postman Collections].

Reviewer: ```suggestion
For an example, see _04 - Event Consumption - Webhook_ > _Webhook Messaging Filtering_ > _Create API_ in the link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee V4 BETA Postman Collection].
```

== Using an asynchronous API

> The link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee Postman Collections] contain several examples of how end users can work with your asynchronous API. Some examples are described on link:event_native_apim_example_use_cases.html#event_consumption[Event consumption].

Reviewer: ```suggestion
The link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee V4 BETA Postman Collection] contains several examples of how end users can work with your asynchronous APIs. Some examples are described on link:event_native_apim_example_use_cases.html#event_consumption[Event consumption].
```

> == Using an asynchronous API

Reviewer: ```suggestion
== Use cases
```
